src/web/ng/impl/WsConnection.hpp

```diff
@@ -0,0 +1,178 @@
+        util::TagDecoratorFactory const& tagDecoratorFactory
+    )
+        requires IsTcpStream<StreamType>
+        : Connection(std::move(ip), std::move(buffer), tagDecoratorFactory)
+        , stream_(std::move(socket))
+        , initialRequest_(std::move(initialRequest))
+    {
+    }
+
+    WsConnection(
+        boost::asio::ip::tcp::socket socket,
+        std::string ip,
+        boost::beast::flat_buffer buffer,
+        boost::asio::ssl::context& sslContext,
+        boost::beast::http::request<boost::beast::http::string_body> initialRequest,
+        util::TagDecoratorFactory const& tagDecoratorFactory
+    )
+        requires IsSslTcpStream<StreamType>
+        : Connection(std::move(ip), std::move(buffer), tagDecoratorFactory)
+        , stream_(std::move(socket), sslContext)
+        , initialRequest_(std::move(initialRequest))
+    {
+        // Disable the timeout. The websocket::stream uses its own timeout settings.
+        boost::beast::get_lowest_layer(stream_).expires_never();
+        stream_.set_option(boost::beast::websocket::stream_base::timeout::suggested(boost::beast::role_type::server));
+        stream_.set_option(
+            boost::beast::websocket::stream_base::decorator([](boost::beast::websocket::response_type& res) {
+                res.set(boost::beast::http::field::server, util::build::getClioFullVersionString());
+            })
+        );
+    }
+
+    std::optional<Error>
+    performHandshake(boost::asio::yield_context yield)
+    {
+        Error error;
+        stream_.async_accept(initialRequest_, yield[error]);
+        if (error)
+            return error;
+        return std::nullopt;
+    }
+
+    bool
+    wasUpgraded() const override
+    {
+        return true;
+    }
+
+    std::optional<Error>
+    send(
+        Response response,
+        boost::asio::yield_context yield,
+        std::chrono::steady_clock::duration timeout = DEFAULT_TIMEOUT
+    ) override
+    {
+        auto error = util::withTimeout(
+            [this, &response](auto&& yield) { stream_.async_write(response.asConstBuffer(), yield); }, yield, timeout
+        );
+        if (error)
+            return error;
+        return std::nullopt;
+    }
+
+    std::expected<Request, Error>
+    receive(boost::asio::yield_context yield, std::chrono::steady_clock::duration timeout = DEFAULT_TIMEOUT) override
+    {
+        auto error = util::withTimeout([this](auto&& yield) { stream_.async_read(buffer_, yield); }, yield, timeout);
+        if (error)
+            return std::unexpected{error};
+
+        auto request = boost::beast::buffers_to_string(buffer_.data());
+        buffer_.consume(buffer_.size());
+
+        return Request{std::move(request), initialRequest_};
+    }
+
+    void
+    close(boost::asio::yield_context yield, std::chrono::steady_clock::duration timeout = DEFAULT_TIMEOUT) override
+    {
+        boost::beast::websocket::stream_base::timeout wsTimeout{};
+        stream_.get_option(wsTimeout);
```

Comment:
do we need timeout here ? seems we don't care about the result of async_close.

Reply:
We don't want this coroutine to hang because client is not responding.